### PR TITLE
Add realm factory helper methods to run work on the correct context

### PR DESCRIPTION
**Pull request by peppy**
Avoids constructing a new `Realm` instance when called from the update thread without worrying about disposal, as discussed [here](https://github.com/realm/realm-dotnet/discussions/2775#discussioncomment-2006438). This avoid blocking the update thread if realm is in a not so great state (by avoiding the implicit `Refresh` call on shared context retrieval).

~~I've updated a variety of usages which would be candidates for causing blocking. We could probably replace all usage of `CreateContext` if everyone is happy with this pattern. I will do this after it's agreed upon. Personally I think this encapsulates realm usages better than what we had.~~

I've updated all usages, apart from `realmFactory.Context` which will come as a follow-up effort.

### CODE REVIEW

**Comment by bdach** [comment]
The code looks fine readability-wise, but if we're talking performance, then what's the delegate allocation overhead of this? Is it enough to justify this change? I guess it's trading off different things, as in update thread blocks caused by fetching the context would be exchanged for potentially longer GCs.

I had briefly considered an alternative of writing an `IDisposable` wrapper about `Realm` that would have two implementations: one for the update thread context, in which `Dispose()` would be a no-op, and one that would create a new context and then dispose of it in `Dispose()`. You could slot in that object where `realmFactory.CreateContext()` is currently used, but it's a less elegant solution, so if the delegate allocs aren't an issue then I'm personally ok with this as is.

**Comment by peppy** [comment]
I also considered that path but honestly at the point of fetching from realm i don't think the overhead of a delegate allocation is of concern (the fetch will outweigh the cost of the delegate in all cases).

**Comment by smoogipoo** [comment]
Personally I'd like to see an `IDisposable` with different behaviour for the update thread. I think I've suggested this a number of already.
But I think this is also fine so long as `CreateContext` is 100% removed. If it can't be removed and there are cases where you have to ping-pong back and forth between using it and this new `Run()` method, then I think an object is required to encapsulate logic more.

**Comment by peppy** [comment]
Would you want to see all usages updated in this PR? There aren't that many, so it should be feasible (unless I come across anything weird).

**Comment by smoogipoo** [comment]
Yes, all usages of `CreateContext` need to be removed if this is to stay.

**Comment by peppy** [comment]
I've made the required changes. Please check each new commit's message for details on why the changes were required. The `RealmLive` change is unavoidable, unfortunately, but should make sense. Let me know if you need me to walk through that.

This PR now obsoletes https://github.com/ppy/osu/pull/16498 since `RealmLive` is using the new methods which allow it to benefit from update thread context usage automatically.

I believe all usage of `realmFactory.Context` can also be updated to this pattern, but will wait for this to be accepted as-is before attempting that.

File: osu.Game.Tests/Database/RealmLiveTests.cs [hunk 1]
```diff
@@ -23,9 +23,9 @@ public void TestLiveEquality()
         {
             RunTestWithRealm((realmFactory, _) =>
             {
-                ILive<BeatmapInfo> beatmap = realmFactory.CreateContext().Write(r => r.Add(new BeatmapInfo(CreateRuleset(), new BeatmapDifficulty(), new BeatmapMetadata()))).ToLive(realmFactory);
+                ILive<BeatmapInfo> beatmap = realmFactory.Run(realm => realm.Write(r => r.Add(new BeatmapInfo(CreateRuleset(), new BeatmapDifficulty(), new BeatmapMetadata()))).ToLive(realmFactory));
```
Comment (frenzibyte) [inline osu.Game.Tests/Database/RealmLiveTests.cs:26]: Could have a `T Write<T>` overload in `RealmContextFactory` and shorten this to:
```csharp
ILive<BeatmapInfo> beatmap = realmFactory.Write(r => r.Add(new BeatmapInfo(CreateRuleset(), new BeatmapDifficulty(), new BeatmapMetadata()))).ToLive(realmFactory);
```

File: osu.Game/Database/RealmContextFactory.cs [hunk 2]
```diff
@@ -72,13 +72,13 @@ public Realm Context
             get
             {
                 if (!ThreadSafety.IsUpdateThread)
-                    throw new InvalidOperationException(@$"Use {nameof(CreateContext)} when performing realm operations from a non-update thread");
+                    throw new InvalidOperationException(@$"Use {nameof(createContext)} when performing realm operations from a non-update thread");
```
Comment (frenzibyte) [inline osu.Game/Database/RealmContextFactory.cs:75]: outdated exception message.
```suggestion
                    throw new InvalidOperationException(@$"Use {nameof(Run)}/{nameof(Write)} when performing realm operations from a non-update thread");
```

File: osu.Game/Overlays/Settings/Sections/Input/KeyBindingsSubsection.cs [hunk 3]
```diff
@@ -34,10 +34,9 @@ private void load(RealmContextFactory realmFactory)
         {
             string rulesetName = Ruleset?.ShortName;
 
-            List<RealmKeyBinding> bindings;
+            List<RealmKeyBinding> bindings = null;
 
-            using (var realm = realmFactory.CreateContext())
-                bindings = realm.All<RealmKeyBinding>().Where(b => b.RulesetName == rulesetName && b.Variant == variant).Detach();
+            realmFactory.Run(realm => bindings = realm.All<RealmKeyBinding>().Where(b => b.RulesetName == rulesetName && b.Variant == variant).Detach());
```
Comment (frenzibyte) [inline osu.Game/Overlays/Settings/Sections/Input/KeyBindingsSubsection.cs:39]: reads better having `bindings` outside the method:
```suggestion
            bindings = realmFactory.Run(realm => realm.All<RealmKeyBinding>().Where(b => b.RulesetName == rulesetName && b.Variant == variant).Detach());
```

**Review by frenzibyte — COMMENTED** [review]
From reviewing code, there are a number of `realmFactory.Run()` usages that could have used `realmFactory.Write()` instead:
 1. `BeatmapManager`'s `Hide()`, `Restore()` and `RestoreAll()` methods.
 2. `EFToRealmMigrator`'s `migrateSkins()` and `migrateSettings()` methods.
 3. `RealmContextFactory`'s `cleanupPendingDeletions()` method (I think?).
 4. `RealmKeyBindingStore`'s `Register()` method.
 5. `KeyBindingRow`'s `updateStoreFromButton` method.
 6. `RealmArchiveModelManager`'s `Delete(TModel)` and `Undelete(TModel)` methods.

Summarized it up here instead of pointing out per-case to reduce noise in this PR.

**Comment by peppy** [comment]
> From reviewing code, there are a number of `realmFactory.Run()` usages that could have used `realmFactory.Write()` instead:

Many were intentionally left because I don't want to unnecessarily start a transaction unless a write is assured. Also I see no immediate need to update every case.

File: osu.Game.Tests/Database/RealmLiveTests.cs [hunk 4]
```diff
@@ -38,22 +38,22 @@ public void TestAccessAfterStorageMigrate()
...
 
-                using (var context = realmFactory.CreateContext())
+                realmFactory.Run(realm =>
                 {
-                    context.Write(r => r.Add(beatmap));
+                    realm.Write(r => r.Add(beatmap));
 
                     liveBeatmap = beatmap.ToLive(realmFactory);
-                }
+                });
```
Comment (bdach) [inline osu.Game.Tests/Database/RealmLiveTests.cs:48]: This test is failing on windows on CI and I suspect this change is the cause. I haven't confirmed experimentally, but my educated guess is that the old test implicitly relied on explicit creation of disposal of a realm context, even though this code was running on the update thread. That was the only realm context used, which means that when it was disposed nothing used realm anymore.

Now with the change to `Run()`, this will use `RealmContextFactory.Context`,  which will not be disposed until the factory is (or until a `BlockAllOperations()` call), causing failures on windows only because it is seems to be the only OS on which the realm file is locked even for reads if any context is open.

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 5]
```diff
@@ -165,8 +171,11 @@ public void RestoreAll()
         /// <returns>A list of available <see cref="BeatmapSetInfo"/>.</returns>
         public List<BeatmapSetInfo> GetAllUsableBeatmapSets()
         {
-            using (var context = contextFactory.CreateContext())
-                return context.All<BeatmapSetInfo>().Where(b => !b.DeletePending).Detach();
+            return contextFactory.Run(realm =>
+            {
+                realm.Refresh();
```
Comment (bdach) [inline osu.Game/Beatmaps/BeatmapManager.cs:176]: What's the rationale for this explicit refresh? Wasn't there before.

Comment (peppy) [reply osu.Game/Beatmaps/BeatmapManager.cs:176]: see https://github.com/ppy/osu/pull/16535/commits/495636538fc23101ec641bdf66eb0ba8b3f6a88c

**Comment by bdach** [comment]
Also, the benchmarks that I merged at #16550 use `CreateContext()`, so those are also going to need to get culled here. Kind of unfortunate.

File: osu.Game.Tests/Database/RealmLiveTests.cs (hunk 1 above)
Comment (peppy) [reply osu.Game.Tests/Database/RealmLiveTests.cs:26]: Considering the only usage of this pattern is in tests, I'll leave it for now. We can add in the future if it becomes a common thing.

File: osu.Game.Tests/Database/RealmLiveTests.cs (hunk 4 above)
Comment (peppy) [reply osu.Game.Tests/Database/RealmLiveTests.cs]: Has been fixed by triggering a recycle 👍

**Review by bdach — APPROVED** [review]
